src/main/scala/org/hammerlab/guacamole/reference/ContigIterator.scala

```diff
@@ -0,0 +1,33 @@
+package org.hammerlab.guacamole.reference
+
+/**
+ * Given some ReferenceRegions, iterate over them as long as they are on @contig.
+ *
+ * Intended for repeated use on the same input-region-iterator, to segment it into runs of reads from the same contig.
+ */
+case class ContigIterator[+R <: ReferenceRegion](contigName: ContigName, regions: BufferedIterator[R])
+  extends BufferedIterator[R] {
+
+  override def head: R =
+    if (hasNext)
+      regions.head
+    else
+      throw new NoSuchElementException
+
+  override def hasNext: Boolean = {
+    regions.hasNext && regions.head.contigName == contigName
+  }
+
+  override def next(): R = {
+    val n = regions.head
```

Comment:
Did you mean `val n = head` instead of `val n = regions.head`? I think as-is it might not throw if the next region has a different contig